src/rest_resources/base.ts

```diff
@@ -0,0 +1,363 @@
+  urlIds: IdSet;
+}
+
+export interface RequestArgs extends BaseFindArgs {
+  http_method: string;
+  operation: string;
+  body?: Body | null;
+  entity?: Base | null;
+}
+
+export interface BaseConstructorArgs {
+  session: SessionInterface;
+  fromData?: Body | null;
+}
+
+interface GetPathArgs {
+  http_method: string;
+  operation: string;
+  urlIds: IdSet;
+  entity?: Base | null;
+}
+
+class Base {
+  // For instance attributes
+  [key: string]: any;
+
+  public static NEXT_PAGE_INFO: GetRequestParams | undefined;
+  public static PREV_PAGE_INFO: GetRequestParams | undefined;
+
+  protected static NAME = '';
+  protected static PLURAL_NAME = '';
+  protected static PRIMARY_KEY = 'id';
+  protected static CUSTOM_PREFIX: string | null = null;
+
+  protected static HAS_ONE: {[attribute: string]: typeof Base} = {};
+  protected static HAS_MANY: {[attribute: string]: typeof Base} = {};
+
+  protected static PATHS: ResourcePath[] = [];
+
+  protected static async baseFind({
+    session,
+    urlIds,
+    params,
+  }: BaseFindArgs): Promise<Base[]> {
+    const response = await this.request({
+      http_method: 'get',
+      operation: 'get',
+      session,
+      urlIds,
+      params,
+    });
+
+    this.NEXT_PAGE_INFO = undefined;
+    this.PREV_PAGE_INFO = undefined;
+    if (response.pageInfo?.nextPage) {
+      this.NEXT_PAGE_INFO = response.pageInfo.nextPage;
+    }
+    if (response.pageInfo?.prevPage) {
+      this.PREV_PAGE_INFO = response.pageInfo.prevPage;
+    }
+
+    return this.createInstancesFromResponse(session, response.body as Body);
+  }
+
+  protected static async request({
+    session,
+    http_method,
+    operation,
+    urlIds,
+    params,
+    body,
+    entity,
+  }: RequestArgs): Promise<RestRequestReturn> {
+    const client = new RestClient(session.shop, session.accessToken);
+
+    const path = this.getPath({http_method, operation, urlIds, entity});
+
+    const cleanParams: {[key: string]: string | number} = {};
+    if (params) {
+      for (const key in params) {
+        if (params[key] !== null) {
+          cleanParams[key] = params[key];
+        }
+      }
+    }
+
+    try {
+      let response: RestRequestReturn | null = null;
+      switch (http_method) {
+        case 'get':
+          response = await client.get({path, query: cleanParams});
+          break;
+        case 'post':
+          response = await client.post({
+            path,
+            query: cleanParams,
+            data: body!,
+            type: DataType.JSON,
+          });
+          break;
+        case 'put':
+          response = await client.put({
+            path,
+            query: cleanParams,
+            data: body!,
+            type: DataType.JSON,
+          });
+          break;
+        case 'delete':
+          response = await client.delete({path, query: cleanParams});
+          break;
+        default:
+          throw new Error('Unrecognized HTTP method');
+      }
+
+      return response;
```

Comment:
1. Could shrink code and improve self-documentation by using `return` in `switch` statement.
2. Error response would be more useful if it stated what the unrecognised HTTP method was

```suggestion
      switch (http_method) {
        case 'get':
          return await client.get({path, query: cleanParams});
        case 'post':
          return await client.post({
            path,
            query: cleanParams,
            data: body!,
            type: DataType.JSON,
          });
        case 'put':
          return await client.put({
            path,
            query: cleanParams,
            data: body!,
            type: DataType.JSON,
          });
        case 'delete':
          return await client.delete({path, query: cleanParams});
        default:
          throw new Error('Unrecognized HTTP method "' + http_method + '"');
      }
```